firmware/port.c

```diff
@@ -707,7 +707,16 @@ void port_handle_extint(PortData *p, u32 flags) {
             if (port_pin_supports_interrupt(p, pin)) {
                 Pin sys_pin = p->port->gpio[pin];
                 if (flags & (1 << pin_extint(sys_pin))) {
-                    port_send_status(p, REPLY_ASYNC_PIN_CHANGE_N + pin);
+                    u8 response = REPLY_ASYNC_PIN_CHANGE_N + pin;
+
+                    // If the pin's state is high, set bit 3
+                    // of the response byte. This will be used to
+                    // inform "change" listeners of the present state
+                    // at the moment of the interrupt.
+                    if (pin_read(port_selected_pin(p))) {
```

Comment:
Yes, but "moment" here is misleading. There is still some delay between
when the interrupt is fired and when the `pin_read` call is executed. It is
still possible (albeit unlikely in most cases) that the state of the pin
changes between those two events. I think it's okay to return these pin
read results because it beats the alternative of doing it in JS but we
should add documentation to state that this "race condition" can exist.

On Mon, Mar 28, 2016 at 9:41 PM Rick Waldron <notifications@github.com>
wrote:

> In firmware/port.c
> <https://github.com/tessel/t2-firmware/pull/158#discussion_r57612383>:
>
> > @@ -707,7 +707,16 @@ void port_handle_extint(PortData *p, u32 flags) {
> >              if (port_pin_supports_interrupt(p, pin)) {
> >                  Pin sys_pin = p->port->gpio[pin];
> >                  if (flags & (1 << pin_extint(sys_pin))) {
> > -                    port_send_status(p, REPLY_ASYNC_PIN_CHANGE_N + pin);
> > +                    u8 response = REPLY_ASYNC_PIN_CHANGE_N + pin;
> > +
> > +                    // If the pin's state is high, set bit 3
> > +                    // of the response byte. This will be used to
> > +                    // inform "change" listeners of the present state
> > +                    // at the moment of the interrupt.
> > +                    if (pin_read(port_selected_pin(p))) {
>
> The reason why we didn't do this on T2 is because of the chance that the
> pin state changes between when the interrupt is fired and the .read call
> occurs
>
> This makes it so that the actual state at the moment of the interrupt is
> captured and no ".read" call (from JS) is necessary.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/tessel/t2-firmware/pull/158/files/37daf453c57c5e2f6ade4cb8313cec3e1add3678#r57612383>
>